### PR TITLE
Xdg config home support

### PR DESCRIPTION
**Pull request by edyounis**
#### Describe your PR, what does it fix/add?

The default config path will now first check if XDG_CONFIG_HOME exists and will use it rather than $HOME/.config if it does exist.

#### Is there anything you want to mention? (unchecked code, possible bugs, found problems, breaking compatibility, etc.)

I didn't read deep enough into the code to understand what OpenGL.cpp is doing, so I am not 100% sure my change makes sense there.

#### Is it ready for merging, or does it need work?

Tested it locally, seems to work in all cases.

Related to #1040

### CODE REVIEW

File: src/config/ConfigManager.cpp [hunk 1]
```diff
@@ -21,22 +21,28 @@ CConfigManager::CConfigManager() {
...
+    configPaths.emplace_back(getMainConfigPath());
 
     Debug::disableLogs = &configValues["debug:disable_logs"].intValue;
     Debug::disableTime = &configValues["debug:disable_time"].intValue;
 
     populateEnvironment();
 }
 
+std::string CConfigManager::getMainConfigPath() {
+    if (g_pCompositor->explicitConfigPath != "")
```
Comment (vaxerski) [inline src/config/ConfigManager.cpp:33]: `.empty()`

File: src/config/ConfigManager.cpp [hunk 2]
```diff
@@ -21,22 +21,28 @@ CConfigManager::CConfigManager() {
...
+        return g_pCompositor->explicitConfigPath;
+    
+    std::string config_home = std::string(getenv("XDG_CONFIG_HOME"));
```
Comment (vaxerski) [inline src/config/ConfigManager.cpp:36]: we use camelCase

+ `std::string(const char*)` requires a nullcheck for the char*

File: src/config/ConfigManager.cpp [hunk 3]
```diff
@@ -21,22 +21,28 @@ CConfigManager::CConfigManager() {
...
+    if (config_home == "")
+        config_home = std::string(getenv("HOME")) + "/.config";
+
+    std::string debug_file_name = std::string("/hypr/hyprlandd.conf");
+    std::string normal_file_name = std::string("/hypr/hyprland.conf");
+    std::string config_file_name = (ISDEBUG ? debug_file_name : normal_file_name);
+    return config_home + config_file_name;
```
Comment (vaxerski) [inline src/config/ConfigManager.cpp:43]: all here as above

File: src/config/ConfigManager.cpp [hunk 4]
```diff
@@ -1297,49 +1301,52 @@ void CConfigManager::loadConfigLoadVars() {
 
     // paths
     configPaths.clear();
+    std::string main_config = getMainConfigPath();
+    Debug::log(LOG, "Using config: %s", main_config.c_str());
+    configPaths.push_back(main_config);
+    std::string config_home = main_config.substr(0, main_config.find_last_of('/'));
+    // find_last_of never returns npos since main_config atleast has /hpyr/
+
+    if (!std::filesystem::is_directory(config_home))
```
Comment (vaxerski) [inline src/config/ConfigManager.cpp:1310]: if () {

Please format your file with the provided `.clang-format`

**Review by vaxerski — CHANGES_REQUESTED** [review]
all of the comments are applicable to the entirety of the mr

File: src/config/ConfigManager.cpp (hunk 2 above)
Comment (Thaodan) [reply src/config/ConfigManager.cpp:36]: Using std::string in that case makes it harder to check if getenv returns nullptr.

**Comment by edyounis** [comment]
Supersedes the superseding: #2107
Implemented the requested changes in the other PR.

File: src/config/ConfigManager.cpp [hunk 5]
```diff
@@ -21,22 +21,31 @@ CConfigManager::CConfigManager() {
...
+    else
+        configPath = xdgConfigHome;
+    return configPath;
+}
+
+std::string CConfigManager::getMainConfigPath() {
+    if (!g_pCompositor->explicitConfigPath.empty())
+        return g_pCompositor->explicitConfigPath;
+
+    return getConfigDir() + (ISDEBUG ? "/hypr/hyprlandd.conf" : "/hypr/hyprland.conf");
```
Comment (fufexan) [inline src/config/ConfigManager.cpp:46]:
```suggestion
    return getConfigDir() + "/hypr/" + (ISDEBUG ? "hyprlandd.conf" : "hyprland.conf");
```

File: src/config/ConfigManager.cpp [hunk 6]
```diff
@@ -1304,49 +1311,49 @@ void CConfigManager::loadConfigLoadVars() {
...
+    configPaths.push_back(mainConfigPath);
+    std::string configPath = mainConfigPath.substr(0, mainConfigPath.find_last_of('/'));
+    // find_last_of never returns npos since main_config at least has /hypr/
+
+    if (!std::filesystem::is_directory(configPath)) {
+        Debug::log(WARN, "Creating config home directory");
+        try {
+            std::filesystem::create_directories(configPath);
+        } catch (...) {
+            parseError = "Broken config file! (Could not create directory)";
```
Comment (fufexan) [inline src/config/ConfigManager.cpp:1325]: Why is the config file broken? It's non-existent. This should probably say `Could not create config directory`.

**Review by fufexan — CHANGES_REQUESTED** [review]

File: src/config/ConfigManager.cpp (hunk 6 above)
Comment (edyounis) [reply src/config/ConfigManager.cpp:1325]: This error message existed as-is before this PR, but will be happy to change it.

**Review by vaxerski — APPROVED** [review]
seems alright to me, wbu @fufexan

**Comment by fufexan** [comment]
all good

**Pull request merged**